source/binding/AssignmentExpressions.cpp

```diff
@@ -548,6 +548,18 @@ Expression& ConversionExpression::fromSyntax(Compilation& compilation,
 }
 
 ConstantValue ConversionExpression::evalImpl(EvalContext& context) const {
+    if (!isImplicit && !type->isCastCompatible(*operand().type, false)) {
```

Comment:
The reason I didn't put it inside static function"convert" is that I need ConversionExpression::isImplicit member variable, unless an extra argument of "convert" is added to pass "isImplicit". The existing "convert" code will return nullptr or assertion fail, if bitstream cast is not processed and returned before, so bitstream cast must be checked first.

An extra isCastCompatible argument and calling isCastCompatible again does not seem elegant. How about adding an extra "bool ConversionExpression::isBitstreamCast" member variable or combine "isImplicit" into an enum, to record what a "ConversionExpression" is?  Doing so may require slight modification of all code calling isCastCompatible.

Reply:
Hmm... I'm not sure why isImplicit matters much here. If you check at the end of convert() for a bitstream cast, then there should be no problem, regardless of whether isImplicit is true or not. It might be some extra code to check for the bitstream cases in convert(), like for example if the target isIntegral() then you need an extra check to see if the source is one of the ones that previous worked (integral, float) or whether it's a bitstream case.

Reply:
It is possible, but there are some complications. The difficult parts are
```c++
    if (to.isUnpackedArray() && from.isUnpackedArray()) {
    if (to.isByteArray()) {
```
For example, for
```sv
   localparam byte a[3:0] = "hello world";
```
Bit-stream casting requires sizes to be the same and need to reject this, but implicit conversion allows truncation and extension, the rules are not exactly the same. For the error message:
```c++
   context.addDiag(diag::ConstEvalDynamicToFixedSize, sourceRange)
```
though bitstreamCast can detect size mismatches too, it is nicer to have different error messages for bit-stream casting and implicit conversions to a user. Implicit conversion allows unpacked arrays of reals, but bit-stream casting forbids it. Of course, we can repeat more type checking on what kind of conversion is in "convert" function, but it defeats the purpose of reusing existing code.
I am thinking adding
```c++
enum class ConversionFlags {
    Implicit = 1,
    BitstreamCast = 2,
    Propagated = 4,
    Invalid = 8
}
BITMASK(ConversionFlags, Invalid)
```
to have more flags for ConversionExpression. In another problem I investigated, the rule in ConstantValue::convertToInt
```c++
    // [11.8.3] says that during an assignment we sign extend iff the rhs is signed.
    // That means we should resize first, and only then change the sign flag if desired.
```
is correct but it is different from the end of [11.8.2] on the same page

> When propagation reaches a simple operand as defined in 11.5, then that operand shall be converted to the propagated type and size. If the operand shall be extended, then it shall be sign-extended only if the propagated type is signed.

and that leads to incorrect evaluation of Slang on an expression
```sv
1'b0 ? 7'd100: 3'sb101
```
where 3'sb101 should have unsigned extension not signed extension. Some flags in ConversionExpression may be needed to handle different conversion rules in different context.

Reply:
Hmm, those are some good points. I'll think about this some more and get back to you.

Reply:
Yeah, I think it makes sense to have this as a separate clause, but I'd like it to be in ::convert so external callers can access it. That could either be via an isImplicit boolean parameter that gets passed in, or an additional enum. I haven't thought too much yet about the potential bug you posted about.

Reply:
Is it ok for me to add a flag in ConversionExpression in addition to isImplicit
```c++
  bool isImplicit;
  bool isBitstreamCast;  // new
```
and create a bitmask\<ConversionFlags\> type to a pass as an optional parameter to ::convert?
With a single bitmask\<ConversionFlags\> parameter, ::convert can accept future enhancements beyond what is needed today. Then from ConversionExpression::evalImpl I can create a bitmask\<ConversionFlags\>  from what ConversionExpression has, pass it to ::convert, and let ::convert do all the work.
With this mechanism, inside ::convert, I can do
```c++
     if (flag & ConversionFlags::BitstreamCast) {
         return to.bitstreamCast(value);
     }
     if ((flag & ConversionFlags::Propagate) && value.isInteger() && to.isIntegral())
        value.integer().setSigned(to.isSigned()); // Fix bug on 1'b0 ? 7'd100: 3'sb101
```

Reply:
It shouldn't be a bitmask right? Just a normal enum? The options are mutually exclusive.

Reply:
Yes, an enum should be sufficient, and a bitmask is not necessary.

Reply:
Ok, that all sounds good to me.